File: Graph/16724번 - 피리 부는 사나이/16724.cpp
#include <iostream>

using namespace std;

int n, m;
char board[1001][1001];
int vis[1001][1001];
int ans = 1;

void input() {
    cin >> n >> m;
    for(int i = 0; i < n; i++) {
        string str;
        cin >> str;
        for(int j = 0; j < str.length(); j++) {
            board[i][j] = str[j];
        }
    }
}

void show() {
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            cout << vis[i][j] << ' ';
        }
        cout << '\n';
    }
    cout << "ans : " << ans << "\n\n";
}

void dfs(int x, int y) {
    vis[x][y] = ans;
    int nx = x;
    int ny = y;
    // show();
    if(board[x][y] == 'U') {
        nx -= 1;
    }
    else if(board[x][y] == 'D') {
        nx += 1;
    }
    else if(board[x][y] == 'L') {
        ny -= 1;
    }
    else if(board[x][y] == 'R'){
        ny += 1;
    }
    
    if(!vis[nx][ny]) {
        dfs(nx, ny);
    }
    else if(vis[nx][ny] == ans) {
        ans++;
    }
    vis[x][y] = vis[nx][ny];
}

void solve() {
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            if(!vis[i][j]) {
                dfs(i, j);
            }
        }
    }
    // show();
    cout << ans - 1 << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Graph/16724번 - 피리 부는 사나이/16724.py
import sys

input = sys.stdin.readline

n, m = map(int, input().split())
board = [input().rstrip() for _ in range(n)]
vis = [[0] * m for _ in range(n)]
ans = 1

def dfs(x, y):
    global ans
    vis[x][y] = ans
    nx, ny = x, y
    if board[x][y] == 'D':
        nx += 1
    elif board[x][y] == 'U':
        nx -= 1
    elif board[x][y] == 'R':
        ny += 1
    else:
        ny -= 1
    
    if vis[nx][ny] == 0:
        dfs(nx, ny)
    elif vis[nx][ny] != 0 and vis[nx][ny] == ans:
        ans += 1
    vis[x][y] = vis[nx][ny]

def solve():
    for i in range(n):
        for j in range(m):
            if vis[i][j] == 0:
                dfs(i, j)
    print(ans - 1)

if __name__ == '__main__':
    solve()